node/src/components/consensus/era_supervisor.rs

```diff
@@ -193,16 +202,48 @@ where
             .iter_past(current_era, bonded_eras * 3)
             .collect();
 
-        info!(?era_ids, "collecting switch blocks");
+        // Asynchronously collect the information needed to initialize all recent eras.
+        let effects = async move {
+            info!(?era_ids, "collecting key blocks");
+            let key_blocks = effect_builder
+                .collect_key_blocks(era_ids.iter().cloned())
+                .await
+                .expect("should have all the key blocks in storage");
 
-        let effects = effect_builder
-            .collect_key_blocks(era_ids)
-            .event(move |key_blocks| Event::InitializeEras {
-                key_blocks: key_blocks.expect("should have all the switch blocks in storage"),
-                validators,
+            if current_era > activation_era_id + bonded_eras * 2 {
+                (key_blocks, None) // All eras can be initialized using the key blocks only.
+            } else {
+                // We need the validator set for the activation era from some protocol state.
+                let state_root_hash = if activation_era_id == current_era {
+                    // The activation era is the current one, so the initial state root hash
+                    // contains its validator set.
+                    initial_state_root_hash
+                } else {
+                    // The parent of the first key block contains the activation era validator set.
+                    *effect_builder
+                        .get_block_at_height_local(key_blocks[&activation_era_id].height() - 1)
```

Comment:
`key_blocks[&activation_era_id].height() - 1` is a switch block for `activation_era_id` – can you extract it to a local variable? It will be easier to read in the future.

Reply:
It's not a switch block; not sure if we should call it a key block — it's just used here to get some state root that we know has the required validator set.

Reply:
What if there was an emergency restart that changed the validator set – it won't in the GS at this height.

Reply:
You're right… I think this actually needs to be:
```suggestion
                        .get_block_at_height_local(key_blocks[&activation_era_id + 1].height() - 1)
```

Reply:
@fizyk20 Can you weigh in? You probably have the best understanding of the emergency restarts.

Reply:
`key_blocks[&activation_era_id]` still _is_ a switch block, though, isn't it? We get it via the `get_switch_block_at_era_id` effect.

I think this could even be `.get_block_at_height_local(key_blocks[&activation_era_id].height() + 1)` - we just need to make sure that the block we use to get the global state hash is _in_ the activation era, and this should guarantee it. Andreas' suggestion should also work, I think.

Reply:
> `key_blocks[&activation_era_id]` still _is_ a switch block, though, isn't it? We get it via the `get_switch_block_at_era_id` effect.

Yes, that's what I thought as well but Andreas seems to disagree.

> 
> I think this could even be `.get_block_at_height_local(key_blocks[&activation_era_id].height() + 1)` - we just need to make sure that the block we use to get the global state hash is _in_ the activation era, and this should guarantee it. Andreas' suggestion should also work, I think.

I prefer your version – adding one era, that Andreas is suggesting, makes it more difficult to see what's going on.

Reply:
> Ah, because we decided that emergency restarts would always happen after a switch block?

Yes, AFAIU an emergency restart will always roll back the history to the beginning of an era.

> Still, it feels a bit safer to not tie this code to even more assumptions, and use only data from after the upgrade/restart.

I don't understand this part – what is (and what is not) tying the code to even more assumptions?

Reply:
I guess you're right. I'll change it.
(The assumption I meant was that that pre-restart block was a switch block, yes.)

Reply:
Ah, and it's a bit uglier because it adds another assumption: If the activation point is era 0, i.e. it's genesis, then there is no such key block.

Reply:
Is this an assumption? It's called _key blocks_ – is it wrong to assume that key block is a switch block? We do that in a couple of places here already.

Reply:
Sorry, I meant another special case.

Reply:
So: The block before the key block for `activation_era_id + 1` works even if the activation era is 0.

If we use the block _after_ the key block for `activation_era_id` instead, we have to add a special case for era 0.

Reply:
> If we use the block after the key block for activation_era_id instead, we have to add a special case for era 0.

Just use `saturating_sub` and it will end on 0.

Reply:
I don't understand.
There is no entry `key_blocks[&0]`, because there is no key block for era 0.

Reply:
Ah, you're right! Sorry. Let's extract it to a method and special case then, as you suggested.

Reply:
> extract it to a method

It's inside an `async` block. If we made it an `async` method of `EraSupervisor`, it would have to borrow a reference to it — we'd have to replace the whole component with an `Arc<Mutex<EraSupervisor>>` which I was hoping to avoid in this PR.

Reply:
I had also tried extracting it to a function (or static method) yesterday, but it took so many arguments that I thought it was better to leave it. Maybe I'll give it another try.

Reply:
:grimacing:

Reply:
> Ah, and it's a bit uglier because it adds another assumption special case: If the activation point is era 0, i.e. it's genesis, then there is no such key block.

Right. This convinces me that Andreas' suggestion is superior and I'd go with that.

Reply:
Are we convinced that auction contract stores validator sets from the past eras though? B/c what you're suggesting requires that – an auction state in era `n` has to know validators from era `n-1`.

Reply:
No, that's why I subtracted 1, so that the block is from the era that we're getting the validators for.

Reply:
The predecessor of the key block for era n is in era n-1 and has the validators for era n-1.

Reply:
Fixed in https://github.com/CasperLabs/casper-node/pull/1023/commits/10ee3692c0bfc436cb953d2f6735070bd4f49f84.